In [84]:
import numpy as np
from matplotlib import pyplot as plt

## $\times$

In [85]:
x = np.array([
    [10, 20],
    [30, 40]
])
y = np.array([
    [1, 2],
    [3, 4]
])

print(x * y)
print(x @ y)
print(np.dot(x, y))
print(np.matmul(x, y))

[[ 10  40]
 [ 90 160]]
[[ 70 100]
 [150 220]]
[[ 70 100]
 [150 220]]
[[ 70 100]
 [150 220]]


In [86]:
print(
    np.array([1, 2]) * np.array([3, 4])
)

[3 8]


In [87]:
x = np.array([
    [
        [0, 1, 2],
        [3, 4, 5]
    ],
    [
        [6, 7, 8],
        [9, 10, 11]
    ],
])
print(x.shape, x)

y = np.array([
    [
        [10, 20],
        [30, 40],
        [50, 60]
    ],
    [
        [50, 60],
        [70, 80],
        [90, 100]
    ]
])
print(y.shape, y)

(2, 2, 3) [[[ 0  1  2]
  [ 3  4  5]]

 [[ 6  7  8]
  [ 9 10 11]]]
(2, 3, 2) [[[ 10  20]
  [ 30  40]
  [ 50  60]]

 [[ 50  60]
  [ 70  80]
  [ 90 100]]]


## x * y
$\begin{align}
x * y =
\begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 1 & 2 \end{bmatrix} \\
        \begin{bmatrix} 3 & 4 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 5 & 6 \end{bmatrix} \\
        \begin{bmatrix} 7 & 8 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
* \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 10 & 20 \end{bmatrix} \\
        \begin{bmatrix} 30 & 40 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 50 & 60 \end{bmatrix} \\
        \begin{bmatrix} 70 & 80 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 1 \times 10 & 2 \times 20 \end{bmatrix} \\
        \begin{bmatrix} 3 \times 30 & 4 \times 40 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 5 \times 50 & 6 \times 60 \end{bmatrix} \\
        \begin{bmatrix} 7 \times 70 & 8 \times 80 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix} \\
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 10 & 40 \end{bmatrix} \\
        \begin{bmatrix} 90 & 160 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 250 & 360 \end{bmatrix} \\
        \begin{bmatrix} 490 & 640 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
\end{align}$

In [88]:
print("x * y = ", x * y)

ValueError: operands could not be broadcast together with shapes (2,2,3) (2,3,2) 

## x @ y (np.matmul())
$\begin{align}
x \@ y =
\begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 1 & 2 \end{bmatrix} \\
        \begin{bmatrix} 3 & 4 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 5 & 6 \end{bmatrix} \\
        \begin{bmatrix} 7 & 8 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
\@ \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 10 & 20 \end{bmatrix} \\
        \begin{bmatrix} 30 & 40 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 50 & 60 \end{bmatrix} \\
        \begin{bmatrix} 70 & 80 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 1 \times 10 + 2 \times 30 & 1 \times 20 + 2 \times 40 \end{bmatrix} \\
        \begin{bmatrix} 3 \times 10 + 4 \times 30 & 3 \times 20 + 4 \times 40 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 5 \times 50 + 6 \times 70 & 5 \times 60 + 6 \times 80 \end{bmatrix} \\
        \begin{bmatrix} 7 \times 50 + 8 \times 70 & 7 \times 60 + 8 \times 80 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix} \\
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix}  70 & 100 \end{bmatrix} \\
        \begin{bmatrix} 150 & 220 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 970 &  780 \end{bmatrix} \\
        \begin{bmatrix} 910 & 1060 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
\end{align}$

In [ ]:
print("x @ y = ", x @ y)
print("np.matmul(x, y) = ", np.matmul(x, y))

## dot product
$\begin{align}
x \cdot y =
\begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 1 & 2 \end{bmatrix} \\
        \begin{bmatrix} 3 & 4 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 5 & 6 \end{bmatrix} \\
        \begin{bmatrix} 7 & 8 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
\cdot \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix} 10 & 20 \end{bmatrix} \\
        \begin{bmatrix} 30 & 40 \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix} 50 & 60 \end{bmatrix} \\
        \begin{bmatrix} 70 & 80 \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix}
            \begin{bmatrix} 1 \times 10 + 2 \times 30 & 1 \times 20 + 2 \times 40 \end{bmatrix} \\
            \begin{bmatrix} 1 \times 50 + 2 \times 70 & 1 \times 60 + 2 \times 80 \end{bmatrix}
        \end{bmatrix} \\
        \begin{bmatrix}
            \begin{bmatrix} 3 \times 10 + 4 \times 30 & 3 \times 20 + 4 \times 40 \end{bmatrix} \\
            \begin{bmatrix} 3 \times 50 + 4 \times 70 & 3 \times 60 + 4 \times 80 \end{bmatrix}
        \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix}
            \begin{bmatrix} 5 \times 10 + 6 \times 30 & 5 \times 20 + 6 \times 40 \end{bmatrix} \\
            \begin{bmatrix} 5 \times 50 + 6 \times 70 & 5 \times 60 + 6 \times 80 \end{bmatrix}
        \end{bmatrix} \\
        \begin{bmatrix}
            \begin{bmatrix} 7 \times 10 + 8 \times 30 & 7 \times 20 + 8 \times 40 \end{bmatrix} \\
            \begin{bmatrix} 7 \times 50 + 8 \times 70 & 7 \times 60 + 8 \times 80 \end{bmatrix}
        \end{bmatrix}
    \end{bmatrix}
\end{bmatrix} \\
&= \begin{bmatrix}
    \begin{bmatrix}
        \begin{bmatrix}
            \begin{bmatrix}  70 & 100 \end{bmatrix} \\
            \begin{bmatrix} 190 & 220 \end{bmatrix}
        \end{bmatrix} \\
        \begin{bmatrix}
            \begin{bmatrix} 150 & 220 \end{bmatrix} \\
            \begin{bmatrix} 430 & 500 \end{bmatrix}
        \end{bmatrix}
    \end{bmatrix} \\
    \begin{bmatrix}
        \begin{bmatrix}
            \begin{bmatrix} 230 & 340 \end{bmatrix} \\
            \begin{bmatrix} 670 & 780 \end{bmatrix}
        \end{bmatrix} \\
        \begin{bmatrix}
            \begin{bmatrix} 310 &  460 \end{bmatrix} \\
            \begin{bmatrix} 910 & 1060 \end{bmatrix}
        \end{bmatrix}
    \end{bmatrix}
\end{bmatrix}
\end{align}$

In [ ]:
z = np.dot(x, y)
print("np.dot(x, y) = ", z.shape, z)